src/sync/sync_session.cpp

```diff
@@ -306,6 +306,13 @@ void SyncSession::handle_error(SyncError error)
     bool should_invalidate_session = error.is_fatal;
     auto error_code = error.error_code;
 
+    if (m_state == &State::dying && error.is_fatal) {
```

Comment:
a. Testing `m_state` without first taking the mutex is racy.
b. Explicitly testing the state doesn't really fit with how the existing state machine code works. It'd be better to add a `handle_error` action to the state machine, which would be a noop for every other state, and could return a boolean indicating whether to consider the error to be handled.